Multi image support

GrygrFlzr opened the pull request:
Requires #2 to be merged first, and then this must be rebased before merging.

Closes #3.

Additional dependency on FxTwitter API, and as such also on the `requests` library.
Attempted to use the builtin `http` library but that ended up with too much logic dedicated to error handling.

In addition to the currently supported nags, additional nag types are now supported:
- Multi-image (bot only posts the mosaic)
- Videos and GIFs (bot only posts the video)
- Two image tweets (bot only posts the second image)

GrygrFlzr committed:
Cut down to direct deps

GrygrFlzr committed:
Multi-image tweet support

Uses FxTwitter API to determine number of images
Supports different nag types
Supports GIF tweets

GrygrFlzr committed:
Don't cast None to string

petzku commented on main.py:
```diff
@@ -22,34 +39,104 @@
     if message.id in nags:
+        guild_string = ""
+        if message.guild is not None:
+            guild_string = f"on {message.guild.id}"
```
This might be cleaner as a ternary, though their "improved" readability is somewhat debatable.
```py
guild_string = f"on {message.guild.id}" if message.guild is not None else ""
```

Also, it might be cleaner to test for boolean truthyness (simply `if message.guild`), as a `Guild` instance will never be falsy. On the other hand, this _is_ more explicit, and we know what The Zen of Python says about that.

GrygrFlzr replied:
Done

petzku commented on main.py:
```diff
@@ -122,12 +216,14 @@ def _is_video_tweet(embed: discord.Embed) -> bool:
     contains_video_thumbnail = "ext_tw_video_thumb" in embed.image.url
-    return contains_video_thumbnail
+    contains_gif_thumbnail = "tweet_video_thumb" in embed.image.url
+    return contains_video_thumbnail or contains_gif_thumbnail
```
Is there a reason this is not one single return statement with both `in` checks inlined?
Or, for that matter, a simple `"_video_thumb/" in embed.image.url`?
Any applicable delimiting slashes are likely good to include as well, whether we have a "full" check or a slightly simpler one.

GrygrFlzr replied:
True, swapped to this

petzku commented on main.py:
```diff
@@ -93,24 +180,31 @@ def should_nag(message: discord.Message) -> bool:
+    tweets: list[tuple[str, str]] = twitter_url_regex.findall(message.content)
+    for user, tid in tweets:
+        # if there's at least one, return the first nag type
+        return get_fx_nagtype(user, tid)
```
Would it not make more sense, if multiple tweets in the message have different "priorities", to pick the highest commonly acceptable one? In a trivial case, a message with two tweet links, the first of which is broken, would result in _no_ response, even if the second one should get a response (say, a mosaic)

GrygrFlzr replied:
Updated logic to match

petzku commented on main.py:
```diff
@@ -93,24 +180,31 @@ def should_nag(message: discord.Message) -> bool:
     3. Embed is safe and there is a video
```
Docstring should probably be updated to reflect the more complex behavior

GrygrFlzr replied:
Updated to match

petzku commented on main.py:
```diff
@@ -93,24 +180,31 @@ def should_nag(message: discord.Message) -> bool:
-            return getattr(channel, "nsfw") is True
+            return NagType.FULL
+        # nsfw in sfw channel, prevent embed
+        print("Skipping NSFW embed in SFW channel")
```
Reminder to myself: move to literally any more sensible logging system

petzku commented on main.py:
```diff
@@ -93,24 +180,31 @@ def should_nag(message: discord.Message) -> bool:
-            return True
-        if hasattr(channel, "nsfw"):
+            return NagType.FULL
+        if hasattr(channel, "nsfw") and getattr(channel, "nsfw") is True:
```
`is True` seems redundant here?

GrygrFlzr replied:
Changed logic to type-narrow using `isinstance`

petzku commented on main.py:
```diff
@@ -122,12 +216,14 @@ def _is_video_tweet(embed: discord.Embed) -> bool:
+    return contains_video_thumbnail or contains_gif_thumbnail
 
 
 def _is_sensitive_tweet_embed(embed: discord.Embed) -> bool:
```
This should possibly also move to using fxtwitter API, but as per Discord conversation, it makes sense ("more likely to be reflective of [discord embed behavior]")

petzku reviewed (changes requested):
`should_nag` behavior in the case of multiple links should be revisited IMO. Otherwise, this looks mostly good to me. Thanks for the PR!

GrygrFlzr committed:
Better type narrowing for nsfw check

GrygrFlzr committed:
Check thread channel nsfw

GrygrFlzr committed:
Proper tweet sensitivity check

GrygrFlzr committed:
Easier video tweet check

GrygrFlzr committed:
Ternary for guild string

GrygrFlzr committed:
Reuse cache on message edits

petzku commented on main.py:
```diff
@@ -83,34 +196,71 @@ def should_spoiler(message: discord.Message) -> bool:
+    # remaining case is POTENTIALLY multi-image, check fx API
+    tweets: list[tuple[str, str]] = twitter_url_regex.findall(message.content)
+
+    # use the most restrictive nag type
```
This could also be done by first resolving the list and picking the maximum level. Caching means this is fast, so likely not an issue, but I dislike the repetition of the iterator.

```py
    nagtypes = [
        get_fx_nagtype(author, tweet_id)
        for author, tweet_id in twitter_url_regex.findall(message.content)
    ]

    for nagtype in (
        None,
        NagType.FULL,
        NagType.MOSAIC,
        NagType.SECOND_IMAGE,
    ):
        if nagtype in nagtypes:
            return nagtype
```

petzku commented on main.py:
```diff
@@ -174,25 +200,34 @@ def should_nag(message: discord.Message) -> Optional[NagType]:
-        if hasattr(channel, "nsfw") and getattr(channel, "nsfw") is True:
-            # .nsfw seems to not exist on some channel types
-            # and discord.py type narrowing is limited
+        if (
+            isinstance(
+                channel,
+                (
+                    discord.TextChannel,
+                    discord.ForumChannel,
+                    discord.StageChannel,
+                    discord.VoiceChannel,
+                    discord.Thread,
+                ),
+            )
+            and channel.is_nsfw()
+        ):
```
Is this purely for type narrowing to "anything that has the is_nsfw property"? It could possibly use protocols: https://peps.python.org/pep-0544/. Implemented in 3.8, so no issue with Python versions either.
```py
class HasNsfwProperty(typing.Protocol):
    def is_nsfw(self) -> bool: ...
```
Would allow this to become just:
```py
        if isinstance(channel, HasNsfwProperty) and channel.is_nsfw():
```

GrygrFlzr replied:
Oh, neat, yeah that was the intention. I think that should be a fine change then.

petzku commented on main.py:
```diff
@@ -201,9 +236,30 @@ def should_nag(message: discord.Message) -> Optional[NagType]:
+    if any(
+        get_fx_nagtype(author, tweet_id) is NagType.VIDEO for author, tweet_id in tweets
+    ):
+        return NagType.VIDEO
```
`get_fx_nagtype` can never return `NagType.VIDEO`, can it?

GrygrFlzr replied:
Currently as implemented, it cannot, but in a future implementation where we decide that Fx is a more reliable way to detect videos/GIFs (vs the current `"_video_thumb/"` string check), it would be a case that might fail.

petzku replied:
Fair enough. Leaving that in, then, but noting this.

petzku reviewed (approved):
Thanks! Looking good; Just a few minor changes remain that I can easily do myself too.

petzku committed:
Use typing.Protocol for type narrowing

This both lets us avoid listing out all the types of channels, as well as being more explicit about _why_ we check for those

petzku committed:
Simplify fxtwitter nag type priority checking
